File: blog/blueprints/blog.py
from flask import Blueprint, request, jsonify
from exts import db, mail
from models import User, Blog



bp = Blueprint("blog", __name__, url_prefix='/blog')

@bp.route('/get_blog', methods=['GET', 'POST'])
def get_blog():
    '''
    期望从前端收到的json文件
    pak = {
        pieces_started_num: (page - 1) *  pieces_of_page,
        pieces_ended_num: page * pieces_of_page
    } 
    '''
    # 获得前端信息
    pieces_started_num = request.json.get('pieces_started_num')
    pieces_ended_num = request.json.get('pieces_ended_num')

    # state 获取数据是否成功
    state = False
    # error 存储返回的错误信息
    error = ''
    # 最后返回的json文件
    result = {}

    # 获得数据库中的blog数据
    try:
        blogs = Blog.query.order_by('deliver_time').all()
        max_length = len(blogs)
        # 获得blogs总数
        blogs_num = max_length
        # 如果下界大于max_length则直接返回
        if int(pieces_started_num) >=max_length:
            return jsonify({'state': False, 'error': 'out of index range!'})
        
        # 如果上界大于max_length则将max_length - 1置为上界
        if int(pieces_ended_num) >= max_length:
            pieces_ended_num = max_length

        # 将blogs按上下界切片
        ret_blogs_list = blogs[int(pieces_started_num): int(pieces_ended_num)]

        # deliver_id 的列表
        deliverer_ids = []
        # post_id的列表
        post_ids = []
        # title列表
        titles = []
        # main_content正文的列表
        main_contents = []
        # tag 列表
        subject_tags = []
        language_tags = []
        mainpoint_tags = []
        deliver_times = []
        # 封面图片路径
        cover_img_paths = []

        for blog in ret_blogs_list:
            deliverer_ids.append(blog.deliverer_id)
            post_ids.append(blog.post_id)
            titles.append(blog.title)
            main_contents.append(blog.main_content)
            subject_tags.append(blog.subject_tag)
            language_tags.append(blog.language_tag)
            mainpoint_tags.append(blog.mainpoint_tag)
            deliver_times.append(blog.deliver_time)
            cover_img_paths.append(blog.cover_img_path)  # 封面图片路径
        
        # 包装result
        result['error'] = error
        result['deliverer_ids'] = deliverer_ids
        result['post_ids'] = post_ids
        result['titles'] = titles
        result['main_contents'] = main_contents
        result['subject_tags'] = subject_tags
        result['language_tags'] = language_tags
        result['mainpoint_tags'] = mainpoint_tags
        result['deliver_times'] = deliver_times
        result['cover_img_path'] = cover_img_paths
        # blog总数
        result['blogs_num'] = blogs_num
        # state置为True
        state = True
        result['state'] = state


        return jsonify(result)

    except Exception as e:
        print(str(e))
        error = str(e)
        result['state'] = state
        result['error'] = error
        return jsonify(result)
    

    

File: blog/js/blog.js
async function main(){    // 获取传递过来参数
    const urlParams = new URLSearchParams(window.location.search)
    const username = urlParams.get('username')
    const userid = urlParams.get('userid')

    // 显示username
    let username_container = document.getElementById('username');
    username_container.innerHTML = username
    getBlogAndShow(1, 6)



    async function getBlogAndShow(page, pieces_of_page = 6) {
        let pak = {
            pieces_started_num: (page - 1) * pieces_of_page,
            pieces_ended_num: page * pieces_of_page
        }
        await fetch('http://127.0.0.1:5001/blog/get_blog', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(pak)
        }).then(response => response.json())
            .then(data => {
                // alert(data.state)
                // 解析后端json
                let deliverer_ids = data.deliverer_ids
                let post_ids = data.post_ids
                let titles = data.titles
                let main_contents = data.main_contents
                let subject_tags = data.subject_tags
                let language_tags = data.language_tags
                let mainpoint_tags = data.mainpoint_tags
                let deliver_times = data.deliver_times
                let cover_img_paths = data.cover_img_path

                // blogs总数
                let blogs_num = data.blogs_num

                // 计算得到pages总数
                let pages_num = (blogs_num / pieces_of_page) >> 0;

                let length = post_ids.length
                // 先清空blog_container
                let old_blog_pieces = document.getElementsByClassName('blog-piece');
                for (let ele of old_blog_pieces){
                    ele.remove();
                }

                // 添加到前端blog_container
                for (let i = 0; i < length; i++){
                    let title = titles[i];
                    let mainContent = main_contents[i];
                    let cover_img_path = cover_img_paths[i];
                    let mainContentMaxLength = mainContent.length;
                    let coverLength = mainContentMaxLength < 300 ? mainContentMaxLength : 300;
                    let coverMainContent = mainContent.slice(0, coverLength); 
                    add_blog(title, coverMainContent, cover_img_path);
                }

                // 添加分页器
                addPagenation(page, pages_num)

            })
            // .catch(err => {
            //     alert(err)
            // })
    }

    function add_blog(title, coverMainContent, cover_img_path){
        let blog_piece = document.createElement('span');
        blog_piece.classList.add("blog-piece");
        // 添加hover特效
        blog_piece.classList.add('mine-shadow-hover');
        // 标题
        let blog_title = document.createElement('span');
        blog_title.classList.add('blog-title');
        blog_title.innerHTML = title;
        // blog-content部分
        let blog_content = document.createElement('span');
        blog_content.classList.add('blog-content');
        // cover-text部分
        let blog_cover_text = document.createElement('span');
        blog_cover_text.classList.add('blog-cover-text');
        blog_cover_text.innerHTML = coverMainContent;
        // cover图片部分
        let blog_cover_img = document.createElement('span');
        blog_cover_img.classList.add('blog-cover-img');
        blog_cover_img.style.backgroundImage = 'url(' + `'` + cover_img_path + `'` +')';
        
        blog_content.appendChild(blog_cover_text);
        blog_content.appendChild(blog_cover_img);

        blog_piece.appendChild(blog_title);
        blog_piece.append(blog_content);

        // 获得blog_container
        let blog_container = document.getElementById('blog-container-1');

        // 加入到blog_container 
        blog_container.appendChild(blog_piece);
    }

    // 加入分页器
    async function addPagenation(page, pages_num){
        let pagenation = document.getElementById('blog-pagenation');
        let page_button_list = document.getElementsByClassName('page_button');
        // 先移除所有page+button
        for (let i = 0; i < page_button_list.length; i++){
            page_button_list[i].remove();
        }
        // 当前页面所在页码
        let standard_page_button;
        // 再加入新的page_button
        alert(pages_num)
        // 获得page_button_container
        let page_button_container = document.getElementById('page_button_container');

        for (let i = 0; i < pages_num; i++){
            // create一个新的page_button
            let page_button = document.createElement('span');
            page_button.classList.add('page_button');
            page_button.innerHTML = i + 1;
            if (i + 1 === page){
                standard_page_button = page_button;
            }

            // 绑定event
            page_button.addEventListener('click', async (event, i) => {
                // d递归调用getBlogAndShow
                getBlogAndShow(i + 1, 6);
            })

            // 加入到page_button_container中
            page_button_container.appendChild(page_button);


        }
        // 置顶到当前页面
        // standard_page_button.scrollIntoView({
        //     behavior: 'smooth' // 平滑滚动
        // });
        
    }

}



main()